lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,247 @@
+	}()
+
+	// Our interceptor makes sure we hold the packet and then signal to the
+	// main loop to handle the packet. We only return true if we were able
+	// to deliver the packet to the main loop.
+	interceptedPacketsChan := make(chan htlcswitch.InterceptedForward)
+	fromSwitchInterceptor := func(packet htlcswitch.InterceptedForward) bool {
+		select {
+		case interceptedPacketsChan <- packet:
+			return true
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+	}
+
+	// Register our interceptor and make sure we resolve the held packets
+	// on disconnect.
+	interceptableForwarder := s.cfg.RouterBackend.InterceptableForwarder
+	interceptableForwarder.SetInterceptor(fromSwitchInterceptor)
+
+	for {
+		select {
+		case intercepted := <-interceptedPacketsChan:
+			log.Tracef("sending intercepted packet to client %v", intercepted)
+			// in case we couldn't forward we exit the loop and drain the
+			// current interceptor as this indicates on a connection problem.
+			if err := rpcInterceptor.holdAndforwardToClient(intercepted); err != nil {
+				return err
+			}
+		case resolution := <-clientResolutionsChan:
+			log.Tracef("resolving intercepted packet %v", resolution)
+			// in case we couldn't resolve we just add a log line since this
+			// does not indicate on any connection problem.
+			if err := rpcInterceptor.resolveFromClient(resolution); err != nil {
+				log.Warnf("client resolution of intercepted "+
+					"packet failed %v", err)
+			}
+		case err := <-errChan:
+			log.Errorf("Received an error: %v, shutting down", err)
+			return err
+		case <-s.quit:
+			return nil
+		}
+	}
+}
+
+func (s *Server) setRPCInterceptor(interceptor *rpcInterceptor) error {
+	s.rpcInterceptorMtx.Lock()
+	defer s.rpcInterceptorMtx.Unlock()
+	if s.currentRPCInterceptor != nil {
+		return ErrInterceptorAlreadyExists
+	}
+	s.currentRPCInterceptor = interceptor
+	return nil
+}
+
+// onInterceptorDisconnected removes all previousely held forwards from
+// the store. Before they are removed it ensure to resume as the default
+// behavior.
+func (s *Server) onInterceptorDisconnected() {
+
+	log.Infof("RPC interceptor disconnected, resolving held packets")
+	interceptableForwarder := s.cfg.RouterBackend.InterceptableForwarder
+
+	// First de-register the current interceptor so we'll stop getting
+	// more interceptions.
+	interceptableForwarder.SetInterceptor(nil)
+
+	// Now drain and resume all previousely forwards and remove the current
+	// interceptor allows another one to connect.
+	s.rpcInterceptorMtx.Lock()
+	interceptor := s.currentRPCInterceptor
+	s.currentRPCInterceptor = nil
+	s.rpcInterceptorMtx.Unlock()
+
+	interceptor.drain()
+}
+
+// rpcInterceptor is a helper struct that handles the lifecycle of an rpc
+// interceptor streaming session.
+// It is created when the stream opens and drains when the stream closes.
+// It should be used from the same go routine and not built for concurrent
+// access.
+type rpcInterceptor struct {
```

Comment:
I would also move this into a separate file. Just easier to navigate

Reply:
I changed the RPC structure, `Server.HtlcInterceptor` only responsible for starting the session, making sure there is only one active `forwardInterceptor` and the interceptor is responsible for the rest.

Reply:
Much improved again